File: StreamProcessorService/src/main/java/com/stream/processor/validation/VitalSignsRangeValidator.java
package com.stream.processor.validation;

import com.stream.processor.Model.SensorEvent;
import org.springframework.stereotype.Component;

@Component
public class VitalSignsRangeValidator implements ValidationRule  {

    private static final double MIN_HEART_RATE = 30.0;
    private static final double MAX_HEART_RATE = 250.0;
    private static final double MIN_OXYGEN_SATURATION = 70.0;
    private static final double MAX_OXYGEN_SATURATION = 100.0;
    private static final double MIN_TEMPERATURE = 35.0;
    private static final double MAX_TEMPERATURE = 42.0;

    public ValidationResult validate(SensorEvent event) {
        ValidationResult result   = ValidationResult.valid();

        if (event.getHeartRate() != null) {
            if (event.getHeartRate() < MIN_HEART_RATE || event.getHeartRate() > MAX_HEART_RATE) {
                result = result.combine(ValidationResult.invalid(
                        String.format("Heart rate out of range: %.1f", event.getHeartRate())
                ));
            }
        }

        if (event.getOxygenSaturation() != null) {
            if (event.getOxygenSaturation() < MIN_OXYGEN_SATURATION ||
                    event.getOxygenSaturation() > MAX_OXYGEN_SATURATION) {
                result = result.combine(ValidationResult.invalid(
                        String.format("Oxygen saturation out of range: %.1f", event.getOxygenSaturation())
                ));
            }
        }

        if (event.getTemperature() != null) {
            if (event.getTemperature() < MIN_TEMPERATURE ||
                    event.getTemperature() > MAX_TEMPERATURE) {
                result = result.combine(ValidationResult.invalid(
                        String.format("Temperature out of range: %.1f", event.getTemperature())
                ));
            }
        }
        return result;
    }
}

File: validation_service/vital_signs_range_validator.py
"""Validates vital signs are within acceptable ranges"""
import sys
import os
from typing import Dict, Any

# Add current directory to path for imports
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from validation_result import ValidationResult


class VitalSignsRangeValidator:
    """Validates vital signs are within acceptable ranges"""
    
    MIN_HEART_RATE = 30.0
    MAX_HEART_RATE = 250.0
    MIN_OXYGEN_SATURATION = 70.0
    MAX_OXYGEN_SATURATION = 100.0
    MIN_TEMPERATURE = 35.0
    MAX_TEMPERATURE = 42.0
    
    def validate(self, event: Dict[str, Any]) -> ValidationResult:
        result = ValidationResult.valid_result()
        
        heart_rate = event.get('heart_rate')
        if heart_rate is not None:
            if heart_rate < self.MIN_HEART_RATE or heart_rate > self.MAX_HEART_RATE:
                result = result.combine(ValidationResult.invalid_result(
                    f"Heart rate out of range: {heart_rate:.1f}"
                ))
        
        oxygen_saturation = event.get('oxygen_saturation')
        if oxygen_saturation is not None:
            if oxygen_saturation < self.MIN_OXYGEN_SATURATION or oxygen_saturation > self.MAX_OXYGEN_SATURATION:
                result = result.combine(ValidationResult.invalid_result(
                    f"Oxygen saturation out of range: {oxygen_saturation:.1f}"
                ))
        
        temperature = event.get('temperature')
        if temperature is not None:
            if temperature < self.MIN_TEMPERATURE or temperature > self.MAX_TEMPERATURE:
                result = result.combine(ValidationResult.invalid_result(
                    f"Temperature out of range: {temperature:.1f}"
                ))
        
        return result

